Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,298 @@
+- `TuringDegree`: The type of Turing degrees, defined as the quotient of partial functions under
+  `TuringEquivalent`.
+
+## Notation
+- `f ≤ᵀ g` : `f` is Turing reducible to `g`.
+- `f ≡ᵀ g` : `f` is Turing equivalent to `g`.
+
+## Implementation Notes
+
+The type of partial functions recursive in an oracle `g` is the smallest type containing
+the constant zero, the successor, projections, the oracle `g`, and is closed under
+pairing, composition, primitive recursion, and μ-recursion.
+
+## References
+* [Carneiro2018] Carneiro, Mario.
+  *Formalizing Computability Theory via Partial Recursive Functions*.
+  arXiv preprint arXiv:1810.08380, 2018.
+* [Odifreddi1989] Odifreddi, Piergiorgio.
+  *Classical Recursion Theory: The Theory of Functions and Sets of Natural Numbers,
+  Vol. I*. Springer-Verlag, 1989.
+* [Soare1987] Soare, Robert I. *Recursively Enumerable Sets and Degrees*. Springer-Verlag, 1987.
+* [Gu2015] Gu, Yi-Zhi. *Turing Degrees*. Institute for Advanced Study, 2015.
+
+## Tags
+Computability, Oracle, Turing Degrees, Reducibility, Equivalence Relation
+-/
+open Primrec Nat.Partrec Part
+
+/-
+This section defines a model of oracle computability and defines
+Turing reducibility and Turing equivalence. We define the Turing Degrees as the quotient under
+Turing equivalence relation
+-/
+
+/--
+The type of partial functions recursive in a set of oracle O is the smallest type containing
+the constant zero, the successor, left and right projections, each oracle g ∈ O, and is closed under
+pairing, composition, primitive recursion, and μ-recursion.
+-/
+
+inductive RecursiveIn (O : Set (ℕ →. ℕ)) : (ℕ →. ℕ) → Prop
+  | zero : RecursiveIn O (fun _ => 0)
+  | succ : RecursiveIn O Nat.succ
+  | left : RecursiveIn O (fun n => (Nat.unpair n).1)
+  | right : RecursiveIn O (fun n => (Nat.unpair n).2)
+  | oracle : ∀ g ∈ O, RecursiveIn O g
+  | pair {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun n => (Nat.pair <$> f n <*> h n))
+  | comp {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun n => h n >>= f)
+  | prec {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y IH => do
+          let i ← IH
+          h (Nat.pair a (Nat.pair y i))
+        )
+      )
+  | rfind {f : ℕ →. ℕ} (hf : RecursiveIn O f) :
+      RecursiveIn O (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n))
+      )
+/--
+f is turing reducible to g if f is recursive in g
+-/
+abbrev TuringReducible (f g : ℕ →. ℕ) : Prop :=
+  RecursiveIn {g} f
+
+/--
+Custom infix notation for `TuringReducible`.
+-/
+infix:50 "≤ᵀ" => TuringReducible
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+abbrev TuringEquivalent (f g : ℕ →. ℕ) : Prop :=
+  AntisymmRel TuringReducible f g
+
+/--
+Custom infix notation for `TuringEquivalent`.
+-/
```

Comment:
This should inherit the docs from `TuringEquivalent` above (see uses of `@[inherit_doc]`)